### PR TITLE
Added basic functionality for Intervals

### PR DESCRIPTION
**Pull request by nishnik**
I have implemented set intervals.
I need some design comments.
I think, I should implement a class `Set` and then inherit `interval` from it, like in logic parallel to `TrigFunction` and `sin`, `cos` etc.
Ping @certik @isuruf

### CODE REVIEW

**Comment by CodeMaxx** [comment]
@nishnik  Add the binary executable to `.gitignore`. That file is not to be committed.

File: symengine/derivative.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include <symengine/polynomial.h>
 #include <symengine/complex_double.h>
 #include <symengine/complex_mpc.h>
+#include <symengine/sets.h>
```
Comment (CodeMaxx) [inline symengine/derivative.cpp]: Is this required in `derivatives.cpp`?

File: symengine/printer.cpp [hunk 2]
```diff
@@ -99,6 +99,20 @@ void StrPrinter::bvisit(const ComplexDouble &x) {
     }
 }
 
+void StrPrinter::bvisit(const Interval &x) {
+    std::ostringstream s;
+    if(x.left_open_)
```
Comment (CodeMaxx) [inline symengine/printer.cpp]: Space after `if`

File: symengine/printer.cpp [hunk 3]
```diff
@@ -99,6 +99,20 @@ void StrPrinter::bvisit(const ComplexDouble &x) {
...
+        s<<"(";
+    else
+        s<<"[";
+    s<<x.start_<<", "<<x.end_;
+    if(x.right_open_)
```
Comment (CodeMaxx) [inline symengine/printer.cpp]: same as above

File: symengine/printer.cpp [hunk 4]
```diff
@@ -99,6 +99,20 @@ void StrPrinter::bvisit(const ComplexDouble &x) {
...
+        s<<")";
+    else
+        s<<"]";
```
Comment (CodeMaxx) [inline symengine/printer.cpp]: Spaces around '<<' in all statements

File: symengine/tests/basic/test_sets.cpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	REQUIRE(r3->is_subset(*r2));
+	REQUIRE(not r3->is_proper_subset(*r2));
+	REQUIRE(not r3->is_proper_superset(*r2));
+
+	r3 = Interval::from_two_nums(*im5, *i20);
+	r4 = Interval::from_two_nums(*zero, *i2);
+	REQUIRE(r3->is_superset(*r4));
+	REQUIRE(r3->is_proper_superset(*r4));
+
+}
```
Comment (CodeMaxx) [inline symengine/tests/basic/test_sets.cpp]: Add newline.

File: symengine/sets.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+		bool is_proper_subset (const Interval &second) const;
+
+		bool is_superset(const Interval &second) const;
+
+		bool is_proper_superset(const Interval &second) const;
+	};
+
+}
+#endif
```
Comment (CodeMaxx) [inline symengine/sets.h]: Add newline

Comment (CodeMaxx) [inline symengine/sets.h]: Newline

File: symengine/sets.cpp [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	RCP<const Interval> Interval::close() const {
+		return Interval::from_mpq(start_, end_, false, false);
+	}
+
+	RCP<const Interval> Interval::interval_intersection(const Interval &second) const {
+		rational_class start, end;
+		bool left_open, right_open;
+		if((this->start_<=second.end_) and (this->end_ >= second.start_)) {
```
Comment (CodeMaxx) [inline symengine/sets.cpp]: Spaces around `<=`

File: symengine/sets.cpp [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	bool Interval::is_superset(const Interval &second) const {
+		return second.is_subset(*this);
+	}
+
+	bool Interval::is_proper_superset(const Interval &second) const {
+		return second.is_subset(*this) and (not __eq__(second));
+	}
+}
```
Comment (CodeMaxx) [inline symengine/sets.cpp]: newline

**Comment by isuruf** [comment]
You need to remove `symengine/tests/basic/test_sets` from the commit history itself as it is a large binary file

**Comment by nishnik** [comment]
@isuruf I need some design comments. Please! look over it.

**Comment by isuruf** [comment]
What is the plan for the intervals? For design decisions, I always look at what SymPy does. Can you point to me the SymPy code?

**Comment by nishnik** [comment]
It is [here](https://github.com/sympy/sympy/blob/master/sympy/sets/sets.py). I talked to Harsh Gupta, he told me that it would be better that we implement `sets` as a super class to `interval`. And from `sets`, we can inherit other types like `emptyset` and others.

**Comment by isuruf** [comment]
There are two options here,

1: A templatized version of interval.
```
template <typename T>
class Interval {
     T left;
     T right;
}
```
    where `T` can be `integer_class`, `rational_class`, `double` or `mpfr_class`

2: Just use `Number` as the type.

CC @hargup, @certik

**Comment by CodeMaxx** [comment]
Can we overload `()` , `[)`, `(]` , `[]` operators rather than passing `true` and `false` for creating every interval?

**Comment by nishnik** [comment]
@isuruf If using this:
>1: A templatized version of interval.
```
template <typename T>
class Interval {
     T left;
     T right;
}
```
where T can be integer_class, rational_class, double or mpfr_class

I think we should use two templates for supporting
```
[0.2, 2] //[rational, integer]
```

**Comment by isuruf** [comment]
That'll quickly become messy with 16 types.

**Comment by nishnik** [comment]
Yes.
I used `rational_class` on the lines of [complex](https://github.com/symengine/symengine/blob/master/symengine/complex.h). It can handle integers too. What is the problem with it? In sympy too, non-real sets are not supported.

**Comment by isuruf** [comment]
Because there are real numbers other than Rationals.

**Comment by nishnik** [comment]
That means `complex` too is not handling irrationals as `real` or `imaginary` part.
I suggest using mpfr_class.

**Comment by isuruf** [comment]
Then you lose the fact that an integer is arbitrary precision because `mpfr_class` has a fixed precision

**Comment by nishnik** [comment]
So, can we use
```
template <typename A, typename B>
class Interval{
     A left;
     B right;
}
```
where `A` and `B` can be `rational_class` or `mpfr_class`.

**Comment by nishnik** [comment]
@isuruf Is the basic structure fine ?

File: symengine/sets.h [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+#include <symengine/basic.h>
+#include <symengine/functions.h>
+#include <symengine/number.h>
+#include <symengine/rational.h>
+#include <symengine/complex.h>
+
+
+namespace SymEngine {
+
+class Interval : public Basic  {
```
Comment (isuruf) [inline symengine/sets.h]: Shouldn't this be derived from `Number`?

Comment (nishnik) [inline symengine/sets.h]: I thought over it while implementing.
But as sympy uses -
`class Set(Basic)`
And in `Number` we have virtual functions like -
`nth_root`, `is_perfect_power`, `pow` etc which don't make sense for a set object.
So, I thought using `Basic` will be better.

Comment (isuruf) [inline symengine/sets.h]: `pow` makes perfect sense.
Think of Interval as a number x s.t. `a < x < b`. Then pow(i, n) would give the range of `x**n`. 

Addition of `a < x < b` and `c < y < d` would give the range of `x+y` which is `a+c, b+d`

Comment (nishnik) [inline symengine/sets.h]: Thanks!
What should I be doing for:
```
    virtual bool is_zero() const = 0;
    virtual bool is_one() const = 0;
    virtual bool is_minus_one() const = 0;
```

Comment (isuruf) [inline symengine/sets.h]: Those should be false

Comment (nishnik) [inline symengine/sets.h]: But @isuruf 
In sympy:
```
In [3]: a = Interval(2,3)

In [4]: b = Interval(5,8)

In [5]: a+b
Out[5]: [2, 3] U [5, 8]

```
And, I think we use that notation of `addition` for fuzzy sets only.

Comment (isuruf) [inline symengine/sets.h]: Hmm, Sage uses the notation I mentioned. http://doc.sagemath.org/html/en/reference/rings_numerical/sage/rings/real_mpfi.html

It depends on what you are trying to represent here.
What are you trying to represent using an Interval

File: symengine/sets.cpp [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+		}
+		else if (eq(*min({s, e}), *e)){
+			throw std::runtime_error("Not a valid Interval");
+		}
+		return true;
+	}
+
+	std::size_t Interval::__hash__() const {
+		std::size_t seed = INTERVAL;
+		hash_combine<long long int>(seed, start_->__hash__());
```
Comment (isuruf) [inline symengine/sets.cpp]: Use *start_

File: symengine/sets.cpp [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+	}
+
+	RCP<const Set> EmptySet::_intersection(const Set &o) const {
+		return emptyset();
+	}
+
+	RCP<const Set> EmptySet::_union(const Set &o) const {
+		if (is_a<Interval>(o)) {
+			const Interval &other = static_cast<const Interval &>(o);
+			return interval(other.start_, other.end_, other.left_open_, other.right_open_);
```
Comment (isuruf) [inline symengine/sets.cpp]: What not return `o`

Comment (nishnik) [inline symengine/sets.cpp]: We have to return `RCP<const Set>` object and `o` is `Set` object

Comment (isuruf) [inline symengine/sets.cpp]: How about taking in a `constRCP<const Set> &o` object?

Comment (certik) [inline symengine/sets.cpp]: If you are returning the argument, then you can take in `const RCP<const Set> &o` and just return it, as @isuruf said.

File: symengine/sets.cpp [hunk 12]
```diff
@@ -0,0 +1,211 @@
+#include <symengine/sets.h>
+#include <symengine/integer.h>
+
+namespace SymEngine {
+
+    Interval::Interval(const RCP<const Number> &start, const RCP<const Number> &end, const bool left_open, const bool right_open)
```
Comment (isuruf) [inline symengine/sets.cpp]: Indentation here is not necessary.

File: symengine/sets.cpp [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+        }
+        else if (eq(*min({s, e}), *e)){
+            throw std::runtime_error("Not a valid Interval");
+        }
+        return true;
+    }
+
+    std::size_t Interval::__hash__() const {
+        std::size_t seed = INTERVAL;
+        hash_combine<long long int>(seed, (*start_).__hash__());
```
Comment (isuruf) [inline symengine/sets.cpp]: See here, https://github.com/symengine/symengine/blob/master/symengine/add.cpp#L57

File: symengine/sets.cpp [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+        }
+
+    bool Interval::is_canonical(const RCP<const Number> &s, const RCP<const Number> &e, bool left_open, bool right_open) const {
+        if (is_a<Complex>(*s) or is_a<Complex>(*e))
+            throw std::runtime_error("Complex set not implemented");
+        if (eq(*e, *s)) {
+            if (left_open or right_open)
+                throw std::runtime_error("Not a valid Interval");
+        }
+        else if (eq(*min({s, e}), *e)){
```
Comment (isuruf) [inline symengine/sets.cpp]: See style guide for formatting if-else statements

File: symengine/sets.cpp [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+                    right_open = other.right_open_;
+                }
+            }
+            else {
+                end = this->end_;
+                right_open = this->right_open_ or other.right_open_;
+            }
+            return interval(start, end, left_open, right_open);
+        }
+        else {
```
Comment (isuruf) [inline symengine/sets.cpp]: See style guide for if-else statements

File: symengine/sets.cpp [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+            return emptyset();
+        }
+    }
+
+    RCP<const Set> Interval::_union(const RCP<const Set> &o) const {
```
Comment (isuruf) [inline symengine/sets.cpp]: Why `_union` instead of `union`?

File: symengine/sets.cpp [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+        SYMENGINE_ASSERT(is_a<Interval>(*o));
```
Comment (isuruf) [inline symengine/sets.cpp]: Don't assume that `o` is a Interval. Use the same logic as `is_subset`

File: symengine/sets.cpp [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+        hash_combine<long long int>(seed, (*end_).__hash__());
+        hash_combine<bool>(seed, left_open_);
+        hash_combine<bool>(seed, right_open_);
+        return seed;
+    }
+
+    bool Interval::__eq__(const Basic &o) const {
+        if (is_a<Interval>(o)) {
+            const Interval &s = static_cast<const Interval&>(o);
+            return ((eq(*this->start_, *s.start_) and eq(*this->end_, *s.end_)) and
```
Comment (isuruf) [inline symengine/sets.cpp]: Remove excess brackets and move the last two bool checks to the front

File: symengine/sets.cpp [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    int Interval::compare(const Basic &o) const {
+        SYMENGINE_ASSERT(is_a<Interval>(o))
+        const Interval &s = static_cast<const Interval &>(o);
+        if ((eq(*this->start_, *s.start_) and eq(*this->end_, *s.end_)) and
+                    ((this->left_open_ == s.left_open_) and (this->right_open_ == s.right_open_)))
+            return 0;
+        else
+            throw std::runtime_error("Not implemented");
```
Comment (isuruf) [inline symengine/sets.cpp]: This shouldn't throw an error

File: symengine/sets.h [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+#include <symengine/functions.h>
+#include <symengine/number.h>
+#include <symengine/rational.h>
+#include <symengine/complex.h>
+
+
+namespace SymEngine {
+
+class Set : public Basic {
+    public:
```
Comment (isuruf) [inline symengine/sets.h]: Indentation here is unnecessary

File: symengine/sets.cpp (hunk 16 above)
Comment (nishnik) [inline symengine/sets.cpp]: `union` is a special class type in C++.

File: symengine/sets.h [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+        virtual bool is_Interval() const = 0;
+        virtual bool is_EmptySet() const = 0;
+        virtual bool is_FiniteSet() const = 0;
+        virtual RCP<const Set> _intersection(const RCP<const Set> &o) const = 0;
+        virtual RCP<const Set> _union(const RCP<const Set> &o) const = 0;
+        virtual bool is_subset(const RCP<const Set> &o) const = 0;
+        virtual bool is_proper_subset (const RCP<const Set> &o) const = 0;
+        virtual bool is_superset(const RCP<const Set> &o) const = 0;
```
Comment (isuruf) [inline symengine/sets.h]: This should just return `not is_proper_subset(o)`

Comment (isuruf) [inline symengine/sets.h]: This should just return `not is_proper_subset(o)`

File: symengine/sets.h [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+        virtual bool is_proper_superset(const RCP<const Set> &o) const = 0;
```
Comment (isuruf) [inline symengine/sets.h]: This should just return `not is_subset(o)`

File: symengine/sets.h (hunk 21 above)
Comment (isuruf) [inline symengine/sets.h]: Ignore my last comment

File: symengine/sets.cpp (hunk 16 above)
Comment (isuruf) [inline symengine/sets.cpp]: Then use the C++ algorithm header notation. `set_union`, `set_intersection`, etc.

File: symengine/sets.cpp [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+        return ((this->left_open_ == s.left_open_) and (this->right_open_ == s.right_open_) and
+                eq(*this->start_, *s.start_) and eq(*this->end_, *s.end_));
+    }
+    return false;
+}
+
+int Interval::compare(const Basic &s) const {
+    // compares two interval based on their length
+    SYMENGINE_ASSERT(is_a<Interval>(s))
+    const Interval &o = static_cast<const Interval &>(s);
```
Comment (isuruf) [inline symengine/sets.cpp]: `compare` doesn't have to make sense. Here, you are doing way more work than necessary.
First compare the bools, `left_open_` and `right_open_` and then compare `start_` and finally `end_`

File: symengine/sets.cpp [hunk 24]
```diff
@@ -0,0 +1,217 @@
...
+bool EmptySet::__eq__(const Basic &o) const {
+    if (is_a<EmptySet>(o))
+        return true;
+    return false;
+}
+
+int EmptySet::compare(const Basic &o) const {
+    if (is_a<EmptySet>(o))
+        return 0;
+    throw std::runtime_error("Not implemented");
```
Comment (isuruf) [inline symengine/sets.cpp]: Checking that `o` is an `EmptySet` should be done only in Debug mode with `SYMENGINE_ASSERT`

File: symengine/sets.cpp (hunk 23 above)
Comment (nishnik) [inline symengine/sets.cpp]: I did not get you.
Could you explain little more ?

Comment (isuruf) [inline symengine/sets.cpp]: Calculating the length is an expensive task. All you have got to do is comparing the bools, and then comparing the numbers.

Comment (nishnik) [inline symengine/sets.cpp]: @isuruf 
Sorry, but I still do not get this.
Can I do the comparison of `this.end_` and `o.start_` and return result accordingly ?

Comment (isuruf) [inline symengine/sets.cpp]: First check `this.left_open_` and `o.left_open_` and then return if they are not equal. Same for `right_open_`. Then do `this.start_->__cmp__(o.start_)` and return if it is not 0. Same for `end_`

Comment (nishnik) [inline symengine/sets.cpp]: @isuruf 
Ping

File: symengine/sets.cpp [hunk 25]
```diff
@@ -0,0 +1,219 @@
...
+    return false;
+}
+
+int Interval::compare(const Basic &s) const {
+    // compares two interval based on their length
+    SYMENGINE_ASSERT(is_a<Interval>(s))
+    const Interval &o = static_cast<const Interval &>(s);
+    if(left_open_ and not o.left_open_) {
+        return -1;
+    } else if (not left_open_ and o.right_open_) {
```
Comment (isuruf) [inline symengine/sets.cpp]: `o.left_open_`

File: symengine/sets.cpp [hunk 26]
```diff
@@ -0,0 +1,219 @@
...
+        return 1;
+    } else {
+        if (right_open_ and not o.right_open_) {
```
Comment (isuruf) [inline symengine/sets.cpp]: Use `else if` here

File: symengine/sets.cpp [hunk 27]
```diff
@@ -0,0 +1,219 @@
...
+            return 1;
+        } else if (not right_open_ and o.right_open_) {
+            return -1;
+        } else {
+            if (start_->__cmp__(*(o.start_)) != 0) {
+                return start_->__cmp__(*(o.start_));
```
Comment (isuruf) [inline symengine/sets.cpp]: Keep the value here in a temporary to avoid re-calculating

File: symengine/sets.cpp (hunk 25 above)
Comment (nishnik) [inline symengine/sets.cpp]: Sorry!
My bad.

**Comment by nishnik** [comment]
Fixed @isuruf 
And Thanks for the review

**Comment by nishnik** [comment]
@isuruf @certik 
Is anything more needed in this PR ?
I plan to implement other types of interval and functions, like `is_disjoint`, `contains` etc in a different PR.

**Comment by certik** [comment]
It looks good to me. If @isuruf is fine with it as well, then we can merge it.

**Comment by CodeMaxx** [comment]
@nishnik You can add some more tests for `Empty Set` . Otherwise +1 .

**Comment by isuruf** [comment]
@CodeMaxx, good point. 
You should add more tests. 
Coverage for `sets.cpp` is 65%. If you enabled coveralls for your repo, then you can see which lines are not covered by the tests.

File: symengine/sets.h [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+    virtual bool is_subset(const RCP<const Set> &o) const;
+    virtual bool is_proper_subset (const RCP<const Set> &o) const;
+    virtual bool is_superset(const RCP<const Set> &o) const;
+    virtual bool is_proper_superset(const RCP<const Set> &o) const;
+};
+
+//! \return RCP<const Set>
+inline RCP<const Interval> interval(const RCP<const Number> &start, const RCP<const Number> &end, const bool left_open = false, const bool right_open = false)
+{
+    return make_rcp<const Interval>(start, end, left_open, right_open);
```
Comment (isuruf) [inline symengine/sets.h]: This should check whether the interval is valid and return an emptyset if not

**Comment by nishnik** [comment]
Ping @isuruf

**Comment by isuruf** [comment]
@nishnik, see here for the coverage report, https://codecov.io/github/symengine/symengine/commit/d60890b63a600821d33c3cf4134e21ad25e0be11

File: symengine/tests/basic/test_sets.cpp [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+    RCP<const Set> r1 = emptyset();
+    RCP<const Set> r2 = interval(zero, one);
+
+    REQUIRE(r1->is_EmptySet());
+    REQUIRE(not r1->is_Interval());
+    REQUIRE(r1->is_FiniteSet());
+    REQUIRE(r1->is_subset(r2));
+    REQUIRE(r1->is_proper_subset(r2));
+    REQUIRE(r1->is_superset(r1));
+    REQUIRE(r1->is_superset(r1));
```
Comment (CodeMaxx) [inline symengine/tests/basic/test_sets.cpp]: This is repeated. I think you meant to use `is_subset`

File: symengine/sets.h [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+ *  \file sets.h
+ *
+ **/
+#ifndef SYMENGINE_SETS_H
+#define SYMENGINE_SETS_H
+
+#include <symengine/basic.h>
+#include <symengine/functions.h>
+#include <symengine/number.h>
+#include <symengine/rational.h>
```
Comment (CodeMaxx) [inline symengine/sets.h]: This is redundant

File: symengine/sets.h [hunk 31]
```diff
@@ -0,0 +1,108 @@
...
+#include <symengine/complex.h>
```
Comment (CodeMaxx) [inline symengine/sets.h]: Redundant

File: symengine/sets.h [hunk 32]
```diff
@@ -34,7 +31,7 @@ class EmptySet : public Set {
     IMPLEMENT_TYPEID(EMPTYSET)
     //EmptySet(EmptySet const&) = delete;
     void operator=(EmptySet const&) = delete;
-    static RCP<const EmptySet> getInstance();
```
Comment (isuruf) [inline symengine/sets.h]: Why this change?

Comment (nishnik) [inline symengine/sets.h]: I did this to generalise that `RCP <const Interval>` or `RCP <const EmptySet>` can only be generated by casting.
See that now, all the functions return `RCP <const Set>`

Comment (isuruf) [inline symengine/sets.h]: I'm not sure I understand. Can you explain a bit more?

Comment (nishnik) [inline symengine/sets.h]: `interval` returns `RCP<const Set>`.
`emptyset` returns `RCP<const Set>`
So, I thought it will be better for `getInstance` to return the same, i.e. `RCP<const Set>` to maintain generality.
Futhermore, returning `RCP<const EmptySet>` won't give us any extra help as all the member functions are virtual.

File: symengine/sets.h [hunk 33]
```diff
@@ -0,0 +1,105 @@
...
+public:
+    RCP<const Number> start_;
+    RCP<const Number> end_;
+    bool left_open_, right_open_;
+public:
+    IMPLEMENT_TYPEID(INTERVAL)
+    virtual std::size_t __hash__() const;
+    virtual bool __eq__(const Basic &o) const;
+    virtual int compare(const Basic &o) const;
+    virtual vec_basic get_args() const { return {}; }
```
Comment (nishnik) [inline symengine/sets.h]: @isuruf 
What should be done here ?

File: symengine/sets.h (hunk 32 above)
Comment (isuruf) [inline symengine/sets.h]: Yes, but the method name is `getInstance` which is supposed to return an instance of `Emptyset`, but it returns a `RCP<const Set>` instead.
Also, if you want an `RCP<const EmptySet>`, you can't get it.

File: symengine/sets.h (hunk 33 above)
Comment (isuruf) [inline symengine/sets.h]: Return `start_` and `end_`

**Comment by nishnik** [comment]
Ping @isuruf !
Thanks for our help.

File: symengine/tests/basic/test_sets.cpp [hunk 34]
```diff
@@ -0,0 +1,135 @@
...
+    r3 = interval(im5, i20);
+    r4 = interval(zero, i2);
+    REQUIRE(r3->is_superset(r4));
+    REQUIRE(r3->is_proper_superset(r4));
+
+    r1 = interval(rat1, rat2);// [5/6, 500/6]
+    r2 = interval(im5, i2); // [-5, 2]
+    r3 = r1->set_intersection(r2);
+    r4 = interval(rat1, i2);
+    REQUIRE(eq(*r3, *r4));
```
Comment (isuruf) [inline symengine/tests/basic/test_sets.cpp]: You can also check that their hashes are equal. In other terms that the hash contract (for any two objects if they are equal, the hashes are equal) is valid.

**Comment by isuruf** [comment]
Looks good to me. Thanks for taking the time to write very thorough test cases.

File: symengine/derivative.cpp (hunk 1 above)
Comment (CodeMaxx) [inline symengine/derivative.cpp]: This can be removed?

File: symengine/derivative.cpp [hunk 35]
```diff
@@ -49,6 +50,8 @@ static RCP<const Basic> diff(const CLASS &self, \
     DIFF0(LeviCivita)
     DIFF0(Max)
     DIFF0(Min)
+    DIFF0(Interval)
```
Comment (CodeMaxx) [inline symengine/derivative.cpp]: Can be removed

File: symengine/derivative.cpp [hunk 36]
```diff
@@ -49,6 +50,8 @@ static RCP<const Basic> diff(const CLASS &self, \
...
+    DIFF0(EmptySet)
```
Comment (CodeMaxx) [inline symengine/derivative.cpp]: Can be removed

Comment (nishnik) [inline symengine/derivative.cpp]: Good Point @CodeMaxx .
Should we return `zero` for their derivatives @isuruf ?

Comment (CodeMaxx) [inline symengine/derivative.cpp]: Follow what sympy does.

Comment (isuruf) [inline symengine/derivative.cpp]: Throwing an error is better. It doesn't make sense for an interval to have a derivative

**Comment by nishnik** [comment]
Ping @isuruf

File: symengine/derivative.cpp [hunk 37]
```diff
@@ -437,6 +438,16 @@ static RCP<const Basic> diff(const CLASS &self, \
                 sub(mul(polygamma(zero, beta_arg1), diff_beta_arg1),
                 mul(polygamma(zero, add(beta_arg0, beta_arg1)), add(diff_beta_arg0, diff_beta_arg1)))));
     }
+
+    static RCP<const Basic> diff(const Interval &self,
+            const RCP<const Symbol> &x) {
+        throw std::runtime_error("Derivative doesn't exist.");
+    }
+
+    static RCP<const Basic> diff(const EmptySet &self,
```
Comment (isuruf) [inline symengine/derivative.cpp]: This can just be `diff(const Set &)`

File: symengine/sets.h [hunk 38]
```diff
@@ -0,0 +1,105 @@
...
+    virtual RCP<const Set> set_union(const RCP<const Set> &o) const;
+    virtual RCP<const Set> set_intersection(const RCP<const Set> &o) const;
+    virtual bool is_subset(const RCP<const Set> &o) const;
+    virtual bool is_proper_subset (const RCP<const Set> &o) const;
+    virtual bool is_superset(const RCP<const Set> &o) const;
+    virtual bool is_proper_superset(const RCP<const Set> &o) const;
+};
+
+//! \return RCP<const Set>
+inline RCP<const Set> emptyset()
```
Comment (isuruf) [inline symengine/sets.h]: This should return a EmptySet

Comment (nishnik) [inline symengine/sets.h]: Fixed!

**Comment by rwst** [comment]
Thanks.
